source/robot_model/src/Model.cpp

```diff
@@ -29,17 +47,145 @@ bool Model::create_urdf_from_string(const std::string& urdf_string, const std::s
+  }
+  std::stringstream buffer;
+  buffer << file_stream.rdbuf();
+  auto urdf = buffer.str();
+
+  pinocchio::urdf::buildModelFromXML(urdf, this->robot_model_);
   this->robot_data_ = pinocchio::Data(this->robot_model_);
+
+  if (load_collision_geometries) {
+    this->init_geom_model(urdf);
+  }
+
   // get the frames
   std::vector<std::string> frames;
   for (auto& f : this->robot_model_.frames) {
     frames.push_back(f.name);
   }
   // remove universe and root_joint frame added by Pinocchio
   this->frames_ = std::vector<std::string>(frames.begin() + 2, frames.end());
-  this->init_qp_solver();
+  this->init_qp_solver();  
+}
+
+// Method to initialize collision geometries
+void Model::init_geom_model(std::string urdf) {
+    try {
+      auto package_paths = this->resolve_package_paths_in_urdf(urdf);
+      pinocchio::urdf::buildGeom(this->robot_model_,
+                                std::istringstream(urdf),
+                                pinocchio::COLLISION, 
+                                this->geom_model_, 
+                                package_paths);
+      this->geom_model_.addAllCollisionPairs();
+    
+      std::vector<pinocchio::CollisionPair> excluded_pairs = this->generate_joint_exclusion_list();
+
+      // remove collision pairs for linked joints (i.e. parent-child joints)
+      for (const auto& pair : excluded_pairs) {
+          this->geom_model_.removeCollisionPair(pair);
+      }
+
+      this->geom_data_ = pinocchio::GeometryData(this->geom_model_);
+    } catch (const std::exception& e) {
+        throw robot_model::exceptions::CollisionGeometryException("Failed to initialize Geomerty model for " + this->get_robot_name());
+    }
+}
+
+std::vector<pinocchio::CollisionPair> Model::generate_joint_exclusion_list() {
+    std::vector<pinocchio::CollisionPair> excluded_pairs;
+    // Iterate through all joints, except the universe joint (0), which has no parent
+    for (pinocchio::JointIndex joint_id = 1u; joint_id < static_cast<pinocchio::JointIndex>(this->robot_model_.njoints); ++joint_id){
+        // Find the parent joint of the current joint
+        pinocchio::JointIndex parent_id = this->robot_model_.parents[joint_id];
+        
+        // TODO: Replace this logic with actual geometry index lookup
+        auto getGeometryIndexForJoint = [](pinocchio::JointIndex joint_id) -> int {
+            return static_cast<int>(joint_id);
+        };
+
+        int geometryIndex1 = getGeometryIndexForJoint(joint_id);
+        int geometryIndex2 = getGeometryIndexForJoint(parent_id);
+
+        // Check if the geometry indices are not equal
+        if (geometryIndex1 != geometryIndex2) {
+            excluded_pairs.push_back(pinocchio::CollisionPair(geometryIndex2, geometryIndex1));
+        }
+    }
+    return excluded_pairs;
+}
+
+unsigned int Model::get_number_of_collision_pairs() {
+  return this->geom_model_.collisionPairs.size();
```

Comment:
What happens if the `geom_model_` is empty constructed if you dont load the collision?

Reply:
In `GeometryModel` in pinocchio is initialized with `CollisionPairVector collisionPairs;` and `CollisionPairVector` is actually `std::vector<CollisionPair>`. After I have check that part of pinocchio, I noticed that I had this unnecessary check initially `if(!this->is_geometry_model_initialized())` but tested that it works perfectly fine without it as expected from the code!

Reply:
as long as this doesn't throw an exception when access properties of a default constructed objects, it's good :+1: